In [1]:
import numpy as np
from astropy.modeling import models
from astropy import coordinates
from astropy import table
from astropy.io import fits, ascii
from asdf import AsdfFile

In [2]:
f = AsdfFile()
print(f.tree)

{}


In [3]:
f.tree['model'] = models.Rotation2D(angle=23)
f.write_to('rotation.asdf')
!less rotation.asdf

#ASDF 1.0.0
#ASDF_STANDARD 1.0.0
%YAML 1.1
%TAG ! tag:stsci.edu:asdf/
--- !core/asdf-1.0.0
asdf_library: !core/software-1.0.0 {author: Space Telescope Science Institute, homepage: 'http://github.com/spacetelescope/pyasdf',
  name: asdf, version: 1.1.0.dev715}
model: !transform/rotate2d-1.0.0 {angle: 23.0}
...


In [4]:
fa = AsdfFile.open('rotation.asdf')
model = fa.tree['model']
print(model(1, 2))

(0.13904259647389283, 2.2317408353941546)


In [5]:
f = AsdfFile()
f.tree['model'] = models.Gaussian1D(amplitude=10, mean=3, stddev=.3)

Validation happens when reading from and writing to a file.

In [6]:
f.write_to("gauss.asdf")

RepresenterError: cannot represent an object: Model: Gaussian1D
Inputs: (u'x',)
Outputs: (u'y',)
Model set size: 1
Parameters:
    amplitude mean stddev
    --------- ---- ------
         10.0  3.0    0.3

In [7]:
a = np.random.rand(2,3)

In [8]:
f = AsdfFile()
f.tree['data1'] = a
f.write_to('array1.asdf')
!less array1.asdf

#ASDF 1.0.0
#ASDF_STANDARD 1.0.0
%YAML 1.1
%TAG ! tag:stsci.edu:asdf/
--- !core/asdf-1.0.0
asdf_library: !core/software-1.0.0 {author: Space Telescope Science Institute, homepage: 'http://github.com/spacetelescope/pyasdf',
  name: asdf, version: 1.1.0.dev715}
data1: !core/ndarray-1.0.0
  source: 0
  datatype: float64
  byteorder: little
  shape: [2, 3]
...
ÓBLK0000¼¨ÙZ“ÀQÜƹ»€þ#Ô>2ˆ
¦å?Dô›E�€Ø?Œ#Z¸+¡ç?lhÔÊ´ÉÍ?,áýMÃcÈ?¤¬¤ÅÙ¯æ?#ASDF BLOCK INDEX
%YAML 1.1
--- [359]
...


In [9]:
f = AsdfFile()
f.tree['data'] = a
f.write_to('array.asdf', auto_inline=20)
!less array.asdf


#ASDF 1.0.0
#ASDF_STANDARD 1.0.0
%YAML 1.1
%TAG ! tag:stsci.edu:asdf/
--- !core/asdf-1.0.0
asdf_library: !core/software-1.0.0 {author: Space Telescope Science Institute, homepage: 'http://github.com/spacetelescope/pyasdf',
  name: asdf, version: 1.1.0.dev715}
data: !core/ndarray-1.0.0
  data:
  - [0.6765201244433805, 0.38284999654553453, 0.7384241676928567]
  - [0.2327180853251919, 0.1905445223635608, 0.7089661464900883]
  datatype: float64
  shape: [2, 3]
...


Arrays that are views into other arrays share the same data in the file.

In [10]:
a1 = a[:1, 2:]
f = AsdfFile()
f.tree['full_array'] = a
f.tree['array_subset'] = a1
f.write_to("shared_array.asdf")#, auto_inline=20)
!less shared_array.asdf
           

#ASDF 1.0.0
#ASDF_STANDARD 1.0.0
%YAML 1.1
%TAG ! tag:stsci.edu:asdf/
--- !core/asdf-1.0.0
array_subset: !core/ndarray-1.0.0
  source: 0
  datatype: float64
  byteorder: little
  shape: [1, 1]
  offset: 16
asdf_library: !core/software-1.0.0 {author: Space Telescope Science Institute, homepage: 'http://github.com/spacetelescope/pyasdf',
  name: asdf, version: 1.1.0.dev715}
full_array: !core/ndarray-1.0.0
  source: 0
  datatype: float64
  byteorder: little
  shape: [2, 3]
...
ÓBLK0000¼¨ÙZ“ÀQÜƹ»€þ#Ô>2ˆ
¦å?Dô›E�€Ø?Œ#Z¸+¡ç?lhÔÊ´ÉÍ?,áýMÃcÈ?¤¬¤ÅÙ¯æ?#ASDF BLOCK INDEX
%YAML 1.1
--- [479]
...


In [11]:
f.write_to('external_array.asdf', all_array_storage='external')
!ls external_array*

external_array.asdf
external_array0000.asdf


In [12]:
!less external_array.asdf

#ASDF 1.0.0
#ASDF_STANDARD 1.0.0
%YAML 1.1
%TAG ! tag:stsci.edu:asdf/
--- !core/asdf-1.0.0
array_subset: !core/ndarray-1.0.0
  source: external_array0000.asdf
  datatype: float64
  byteorder: little
  shape: [1, 1]
  offset: 16
asdf_library: !core/software-1.0.0 {author: Space Telescope Science Institute, homepage: 'http://github.com/spacetelescope/pyasdf',
  name: asdf, version: 1.1.0.dev715}
full_array: !core/ndarray-1.0.0
  source: external_array0000.asdf
  datatype: float64
  byteorder: little
  shape: [2, 3]
...


In [13]:
!less external_array0000.asdf

#ASDF 1.0.0
#ASDF_STANDARD 1.0.0
%YAML 1.1
%TAG ! tag:stsci.edu:asdf/
--- !core/asdf-1.0.0
asdf_library: !core/software-1.0.0 {author: Space Telescope Science Institute, homepage: 'http://github.com/spacetelescope/pyasdf',
  name: asdf, version: 1.1.0.dev715}
...
ÓBLK0000¼¨ÙZ“ÀQÜƹ»€þ#Ô>2ˆ
¦å?Dô›E�€Ø?Œ#Z¸+¡ç?lhÔÊ´ÉÍ?,áýMÃcÈ?¤¬¤ÅÙ¯æ?#ASDF BLOCK INDEX
%YAML 1.1
--- [264]
...


Opening a file in exploded form collects all parts and recreates the original tree.

In [14]:
fa=AsdfFile.open('external_array.asdf')

print fa.tree['array_subset']#.copy()

<array (unloaded) shape: [1, 1] dtype: float64>


In [15]:
fa.add_history_entry("Source CDP-5", {'name': "jreftools", "author": 'STScI',
                                       "homepage": "http://github.com/spacetelescope/jreftools",
                                       "version": "0.6"})
fa.write_to("with_history.asdf")
!less with_history.asdf

#ASDF 1.0.0
#ASDF_STANDARD 1.0.0
%YAML 1.1
%TAG ! tag:stsci.edu:asdf/
--- !core/asdf-1.0.0
array_subset: !core/ndarray-1.0.0
  source: with_history0000.asdf
  datatype: float64
  byteorder: little
  shape: [1, 1]
  offset: 16
asdf_library: !core/software-1.0.0 {author: Space Telescope Science Institute, homepage: 'http://github.com/spacetelescope/pyasdf',
  name: asdf, version: 1.1.0.dev715}
full_array: !core/ndarray-1.0.0
  source: with_history0000.asdf
  datatype: float64
  byteorder: little
  shape: [2, 3]
history:
- !core/history_entry-1.0.0
  description: Source CDP-5
  software: !core/software-1.0.0 {author: STScI, homepage: 'http://github.com/spacetelescope/jreftools',
    name: jreftools, version: '0.6'}
  time: 2016-04-14 16:30:43.100000
...


Schemas available in asdf 1.1

https://github.com/spacetelescope/asdf-standard/tree/master/schemas/stsci.edu/asdf

Adding a new schema to ASDF

- Write the schema in yaml and add it to asdf-standard
- Write the Type class implementing two mehtods - "from_tree" and "to_tree" and add the class to asdf/tags